File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
 /*
 1. create a set, charSet
 res = 0
 l = 0
 2. r = 0 to s.length {
    while (s[r] in set) {
        remove s[l] from set
        charSet.delete(s[l])
        l++
    }
    Add s[r] to set
    res = max(res, r - l + 1)
 }
 3. return res
 */
var lengthOfLongestSubstring = function(s) {
    var charSet = new Set()
    var l = 0
    var res = 0

    for (var r = 0; r < s.length; r++) {
        while (charSet.has(s[r])) {
            charSet.delete(s[l])
            l++
        }
        charSet.add(s[r])
        res = Math.max(res, r - l + 1)
    }
    return res
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
'''
1. l = 0
r = 0
charSet = set to track seen characters
res = 0
2. 
while r < len(string):
    while s[r] is seen:
        charSet.remove(s[l])
        l += 1
    res = max(res, r - l + 1)
    add s[r] in set
'''

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        charSet = set()
        l = 0
        r = 0

        for r in range(0, len(s)):
            while s[r] in charSet:
                charSet.remove(s[l])
                l+=1
            
            res = max(res, r - l + 1)

            charSet.add(s[r])

        return res
            